### PR TITLE
support procs in asset_path (issue #9)

### PR DESCRIPTION
**Pull request by ryanong**

### CODE REVIEW

**Comment by ryanong** [comment]
fixed tests

**Comment by ryanong** [comment]
@sikachu is this okay?

**Comment by sikachu** [comment]
Just merged it in (like a year too late). Thanks for your patch!

**Comment by sikachu** [comment]
Actually, I found some issue after reading the code through a bit more. I just reverted this in 84f0fa7ceeddd384a00e29deb43fee53980dff70 and going to comment in the PR on the issues I found.

File: lib/sprockets/redirect.rb [hunk 1]
```diff
@@ -74,14 +76,36 @@ def digest_path
 
     # Sends a redirect header back to browser
     def redirect_to_digest_version(env)
-      url = URI(@asset_host || @request.url)
-      url.path = "#{@prefix}/#{digest_path}"
+      url = File.join(compute_asset_host, @prefix, digest_path)
```
Comment (sikachu) [inline lib/sprockets/redirect.rb]: `File.join` is not appropriate for URI, as this will join the URL with `\` if for some reason this is being run on Windows. Would you mind changing this to [`URI.join`](http://ruby-doc.org/stdlib-2.0.0/libdoc/uri/rdoc/URI.html#method-c-join)?

File: lib/sprockets/redirect.rb [hunk 2]
```diff
@@ -74,14 +76,36 @@ def digest_path
...
-      [self.class.redirect_status, headers, [redirect_message(url.to_s)]]
+
+      [self.class.redirect_status, headers, [redirect_message(url)]]
+    end
+
+    def compute_asset_host
+      host = @asset_host
+
+      if host.respond_to?(:call)
+        arity = host.respond_to?(:arity) ? host.arity : host.method(:call).arity
```
Comment (sikachu) [inline lib/sprockets/redirect.rb]: This logic is very complex.

I believe we don't need to check for arity. Just call with `logical_path, request`. User can ignore those in the lambda, or it would automatically get ignored if they use Proc.

File: lib/sprockets/redirect.rb [hunk 3]
```diff
@@ -74,14 +76,36 @@ def digest_path
...
+        args = [logical_path]
+        args << @request if @request && (arity > 1 || arity < 0)
+        host = host.call(*args)
+      elsif host =~ /%d/
```
Comment (sikachu) [inline lib/sprockets/redirect.rb]: I think this section is unrelated to this patch.

File: lib/sprockets/redirect.rb [hunk 4]
```diff
@@ -74,14 +76,36 @@ def digest_path
...
+        host = host % (Zlib.crc32(source) % 4)
+      end
+
+      host ||= @request.base_url if @request
```
Comment (sikachu) [inline lib/sprockets/redirect.rb]: This `if` statement shouldn't be necessary. `@request` should always be here, right?

File: lib/sprockets/redirect.rb [hunk 5]
```diff
@@ -74,14 +76,36 @@ def digest_path
...
+      return unless host
```
Comment (sikachu) [inline lib/sprockets/redirect.rb]: I think the cause below can be removed altogether.

If user return something from Proc that's not a URL with scheme it's pretty much their bug and we shouldn't try to safeguard them.

I also feel like maybe this whole method should return an URI object so that you can just call `join` on line 79.

**Comment by sikachu** [comment]
I updated your patch in db6dda52cdddbcd80f3b98140dca8f8e37ddc9b7 on top of your commit and going to merge that in soon after I finish with the README. I think this can once again be closed.

Thanks again for your original patch.
